### PR TITLE
Implement Moyal distribution

### PR DESCRIPTION
**Pull request by fearghuskeeble**
This adds a [Moyal distribution](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.moyal.html). Since this is pretty similar to the Gumbel I followed the recipe from that.

### CODE REVIEW

File: tensorflow_probability/python/bijectors/moyal_cdf.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
+# Copyright 2018 The TensorFlow Probability Authors.
```
Comment (srvasude) [inline tensorflow_probability/python/bijectors/moyal_cdf.py:1]: 2020

File: tensorflow_probability/python/bijectors/moyal_cdf.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+  def _inverse(self, y):
+    with tf.control_dependencies(self._maybe_assert_valid_y(y)):
+      return (self.loc - self.scale *
+              tf.math.log(2 * tf.math.square(tf.math.erfinv(1 - y))))
+
+  def _inverse_log_det_jacobian(self, y):
+    with tf.control_dependencies(self._maybe_assert_valid_y(y)):
+      return (tf.math.square(tf.math.erfinv(1 - y)) +
+              tf.math.log(self.scale * np.sqrt(np.pi)) -
```
Comment (srvasude) [inline tensorflow_probability/python/bijectors/moyal_cdf.py:112]: Can separate out the `np.sqrt` to be `0.5 * np.log(np.pi)`

File: tensorflow_probability/python/bijectors/moyal_cdf.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+              tf.math.log(tf.math.erfinv(1 - y)))
+
+  def _forward_log_det_jacobian(self, x):
+    scale = tf.convert_to_tensor(self.scale)
+    z = (x - self.loc) / scale
+    return -1/2 * (tf.exp(-z) + z + tf.math.log(2 * np.pi * tf.square(scale)))
```
Comment (srvasude) [inline tensorflow_probability/python/bijectors/moyal_cdf.py:118]:
```suggestion
    return -0.5 * (tf.exp(-z) + z + np.log(2 * np.pi) + 2. * tf.math.log(scale))
```

File: tensorflow_probability/python/bijectors/moyal_cdf_test.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
+# Copyright 2018 The TensorFlow Probability Authors.
```
Comment (srvasude) [inline tensorflow_probability/python/bijectors/moyal_cdf_test.py:1]: 2020

File: tensorflow_probability/python/distributions/__init__.py [hunk 5]
```diff
@@ -205,6 +206,7 @@
     'Triangular',
     'TruncatedNormal',
     'Uniform',
+    'Moyal',
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/__init__.py:209]: Move up?

File: tensorflow_probability/python/distributions/moyal.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
+# Copyright 2018 The TensorFlow Probability Authors.
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/moyal.py:1]: 2020

File: tensorflow_probability/python/bijectors/moyal_cdf.py (hunk 3 above)
Comment (srvasude) [inline tensorflow_probability/python/bijectors/moyal_cdf.py:118]: Prefer 0.5 vs. 1/2. Here and elsewhere

File: tensorflow_probability/python/distributions/moyal.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+    # TODO(b/145620027) Finalize choice of bijector. Consider switching to
+    # Chain([Softplus(), Log()]) to lighten the doubly-exponential right tail.
+    return identity_bijector.Identity(validate_args=self.validate_args)
+
+  def _parameter_control_dependencies(self, is_init):
+    return self._moyal_bijector._parameter_control_dependencies(is_init)  # pylint: disable=protected-access
+
+
+@kullback_leibler.RegisterKL(Moyal, Moyal)
+def _kl_moyal_moyal(a, b, name=None):
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/moyal.py:218]: Can you add a reference for this?

Comment (fearghuskeeble) [reply tensorflow_probability/python/distributions/moyal.py:218]: I solved this with Mathematica, haven't found a reference as this distribution is not exactly widely used...

File: tensorflow_probability/python/distributions/moyal_test.py [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+    # This reshape is intended to expand the number of test cases.
+    a_loc = a_loc.reshape((len(a_loc), 1, 1, 1))
+    a_scale = a_scale.reshape((1, len(a_scale), 1, 1))
+    b_loc = b_loc.reshape((1, 1, len(b_loc), 1))
+    b_scale = b_scale.reshape((1, 1, 1, len(b_scale)))
+
+    a = tfd.Moyal(loc=a_loc, scale=a_scale, validate_args=True)
+    b = tfd.Moyal(loc=b_loc, scale=b_scale, validate_args=True)
+
+    true_kl = (0.5 * (np.power(2, a_scale/b_scale) *
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/moyal_test.py:265]: I might remove the true_kl computation. The computation itself is a bit complicated and I think the checking sample statistics (i.e. sample KL is consistent with analytic KL) is probably good enough.

**Comment by fearghuskeeble** [comment]
@srvasude The new tests seem to pass now - I just fixed a bijector one in this commit.

File: tensorflow_probability/python/distributions/moyal.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+from tensorflow_probability.python.internal import dtype_util
+from tensorflow_probability.python.internal import tensor_util
+
+__all__ = [
+    'Moyal',
+]
+
+
+class Moyal(transformed_distribution.TransformedDistribution):
+  """The scalar Moyal distribution with location `loc` and `scale` parameters.
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/moyal.py:41]: Is there a vector Moyal distribution? Maybe drop 'scalar'?

File: tensorflow_probability/python/distributions/moyal.py (hunk 7 above)
Comment (srvasude) [reply tensorflow_probability/python/distributions/moyal.py:219]: Can you add an inline comment specifying that (that this was computed via Mathematica)?

**Pull request merged**